File: code.eval/01-hard/poker_hands/poker.hands.py
# Reuben Thorpe (2016), CodeEval [Poker Hands v1.0]
from collections import defaultdict
from collections import Counter
from copy import deepcopy
from sys import argv


class Hand:
    """
        An abstract class representing a poker hand with various helper
        functions such as "is a hand sequential?".
    """

    def __init__(self, cards):
        self.cards = self._get_type_map(cards)
        self.numbers = self._get_card_numbers()
        self.types = len(self.cards)
        self.groups = list()
        self.counter = Counter()
        self.hand_map = {'High Card': 0, 'One Pair': 1,
                         'Two Pairs': 2, 'Three of a Kind': 3,
                         'Straight': 4, 'Flush': 5,
                         'Full House': 6, 'Four of a Kind': 7,
                         'Straight Flush': 8, 'Royal Flush': 9}


    def _get_type_map(self, cards):
        """
            Map non numeric identifiers into numbers, also organise cards by
            type.
        """
        name_map = {'T': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14}
        hand_map = defaultdict(set)

        for card in cards:
            num, colour = card[0], card[1]

            if num.isdigit():
                hand_map[colour].add(int(num))

            else:
                hand_map[colour].add(name_map[num])

        return(hand_map)


    def _get_card_numbers(self):
        """
            Get all the card numbers from the hand, irespective of type.
        """
        return([num for colr in self.cards.values() for num in colr])


    def min(self):
        """
            Get minimum card number from hand.
        """
        return(min(self.numbers))


    def max(self):
        """
            Get maximum card number from hand.
        """
        return(max(self.numbers))


    def is_sequential(self):
        """
            Test whether the values in a hand are in sequential order.
        """
        card_numbs = set(deepcopy(self.numbers))


        if len(card_numbs) == 5:
            transient = min(card_numbs)
            card_numbs.remove(transient)

            while card_numbs and min(card_numbs) == (transient + 1):
                transient = min(card_numbs)
                card_numbs.remove(transient)

            return(False if card_numbs else True)

        else:
            return(False)


    def n_of_a_kind(self):
        """
            Container for grouping information. How many pairs are there?
        """
        if not self.groups:
            self.counter = Counter(self.numbers)
            self.groups = [i for i in self.counter.values() if i > 1]

        return(self.groups)


    def highest_hand(self):
        """
            Compute the highest possible hand.
        """
        hand_map = self.hand_map

        if self.types == 1:
            if self.is_sequential():
                if self.min() == 10:
                    return(hand_map['Royal Flush'], None)
                else:
                    return(hand_map['Straight Flush'], sum(self.numbers))

            else:
                return(hand_map['Flush'], sum(self.numbers))


        elif self.types == 4 and 4 in self.n_of_a_kind():
            weight = 4 * self.counter.most_common(1)[0][0]
            return(hand_map['Four of a Kind'], weight)


        elif self.is_sequential():
            return(hand_map['Straight'], sum(self.numbers))


        elif self.types >= 3 and 3 in self.n_of_a_kind():
            if 2 in self.n_of_a_kind():
                return(hand_map['Full House'], sum(self.numbers))
            else:
                weight = 3 * self.counter.most_common(1)[0][0]
                return(hand_map['Three of a Kind'], weight)


        elif self.types >= 2 and 2 in self.n_of_a_kind():
            if self.n_of_a_kind().count(2) == 2:
                weight = sum(2*num for num, occ in self.counter.most_common(2))
                return(hand_map['Two Pairs'], weight)
            else:
                weight = 2 * self.counter.most_common(1)[0][0]
                return(hand_map['One Pair'], weight)

        else:
            return(hand_map['High Card'], self.max())


def compare_poker_hands(left_hand, right_hand):
    """
        Compare two poker hand objects and return the one with the highest
        hand.
    """

    l_highest_hand, l_weight = left_hand.highest_hand()
    r_highest_hand, r_weight = right_hand.highest_hand()

    if l_highest_hand > r_highest_hand:
        return('left')

    elif r_highest_hand > l_highest_hand:
        return('right')

    else:
        if l_weight == None:
            return('none')

        elif l_weight > r_weight:
            return('left')

        elif r_weight > l_weight:
            return('right')

        else:
            left_numbers = left_hand.numbers
            right_numbers = right_hand.numbers

            while left_numbers:
                max_l = max(left_numbers)
                max_r = max(right_numbers)

                if max_l > max_r:
                    return('left')

                elif max_r > max_l:
                    return('right')

                else:
                    left_numbers.remove(max_l)
                    right_numbers.remove(max_r)

            return('none')


def parse_problem(file_path):
    """
        Parse the CodeEval problem "Poker Hands".
    """
    with open(file_path, 'r') as in_file:
        for line in in_file:
            line = line.strip()
            if line:
                cards = line.split()
                left_hand = Hand(cards[0:5])
                right_hand = Hand(cards[5:])

                yield(left_hand, right_hand)


if __name__ == '__main__':
    for problem in parse_problem(argv[1]):
        left_hand, right_hand = problem
        print(compare_poker_hands(left_hand, right_hand))

File: code.eval/01-hard/poker_hands/poker.hands.c
// Reuben Thorpe (2016), CodeEval [Poker Hands v1.1]
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdbool.h>


void get_highest_score_in_hand(int hand[5][2], int score[2], int num_seq[5]);
void n_of_a_kind(int num_seq[5], int n_of_kind[2], int* n_of_kind_weight);
const char* compare_poker_hands(int l_hand[5][2], int r_hand[5][2]);
void get_hands(FILE* in_file, int l_hand[5][2], int r_hand[5][2]);
void convert_to_numeric(char parse_hand[5][3], int hand[5][2]);
void get_num_sequence(int hand[5][2], int num_seq[5]);
void count_types_in_hand(int hand[5][2], int* types);
int cmpfunc (const void * a, const void * b);
bool is_hand_sequential(int num_seq[5]);
int get_max_id(int num_seq[5]);
int name_to_num(char name);
int min(int num_seq[5]);
int max(int num_seq[5]);
int sum(int num_seq[5]);


int main (int argc, const char* argv[]) {
  FILE* in_file = fopen(argv[1], "r");
  int l_hand[5][2], r_hand[5][2];

  while (!feof(in_file)) {
    get_hands(in_file, l_hand, r_hand);
    printf("%s\n", compare_poker_hands(l_hand, r_hand));
  }
  return(0);
}


const char* compare_poker_hands(int l_hand[5][2], int r_hand[5][2]) {
  // Comapare two poker hands and decide which one wins!
  int l_score[2], r_score[2], l_num_seq[5], r_num_seq[5], types;

  int max_l, max_r;

  get_highest_score_in_hand(l_hand, l_score, l_num_seq);
  get_highest_score_in_hand(r_hand, r_score, r_num_seq);

  if (l_score[0] > r_score[0]) return("left");

  else if (r_score[0] > l_score[0]) return("right");

  else {

    if (l_score[1] > r_score[1]) return("left");

    else if (r_score[1] > l_score[1]) return("right");

    else {

      while (max(l_num_seq) != 0) {
        max_l = max(l_num_seq);
        max_r = max(r_num_seq);

        if (max_l > max_r) return("left");

        else if (max_r > max_l) return("right");

        else {
          l_num_seq[get_max_id(l_num_seq)] = 0;
          r_num_seq[get_max_id(r_num_seq)] = 0;
        }
      }
      return("none");
    }
  }
}


void get_highest_score_in_hand(int hand[5][2], int score[2], int num_seq[5]) {
  // Generate the highest achivable score by a poker hand.
  int types, n_of_kind_weight, n_of_kind[2] = {0};

  get_num_sequence(hand, num_seq);
  count_types_in_hand(hand, &types);
  n_of_a_kind(num_seq, n_of_kind, &n_of_kind_weight);


  if (types == 1) {
    if (is_hand_sequential(num_seq)) {
      if (min(num_seq) == 10) {
        // Royal Flush
        score[0] = 9;
        score[1] = 0;
        return;
      }
      else {
        // Straight Flush
        score[0] = 8;
        score[1] = sum(num_seq);
        return;
      }
    }
    else {
      // Flush
      score[0] = 5;
      score[1] = sum(num_seq);
      return;
    }
  }

  else if (types == 4 && (n_of_kind[0] == 4 || n_of_kind[1] == 4)) {
    // Four of a Kind
    score[0] = 7;
    score[1] = n_of_kind_weight;
    return;
  }

  else if (is_hand_sequential(num_seq)) {
    // Straight
    score[0] = 4;
    score[1] = sum(num_seq);
  }

  else if (types >= 3 && (n_of_kind[0] == 3 || n_of_kind[1] == 3)) {
    if (n_of_kind[0] == 2 || n_of_kind[1] == 2) {
      // Full House
      score[0] = 6;
      score[1] = n_of_kind_weight;
      return;
    }
    else {
      // Three of a Kind 3
      score[0] = 3;
      score[1] = n_of_kind_weight;
      return;
    }
  }

  else if (types >= 2 && (n_of_kind[0] == 2 || n_of_kind[1] == 2)) {
    if (n_of_kind[0] == 2 && n_of_kind[1] == 2) {
      // Two Pairs
      score[0] = 2;
      score[1] = n_of_kind_weight;
      return;
    }
    else {
      // One Pair
      score[0] = 1;
      score[1] = n_of_kind_weight;
      return;
    }
  }

  else {
    score[0] = 0;
    score[1] = max(num_seq);
    return;
  }

}


void get_num_sequence(int hand[5][2], int num_seq[5]) {
  // Get all numeric values of cards in a hand disregarding type.
  for (int i = 0; i < 5; i++) num_seq[i] = hand[i][0];
  qsort(num_seq, 5, sizeof(int), cmpfunc);
}


int min(int num_seq[5]) {
  // Get the minimum from a numeric list.
  int min = num_seq[0];

  for (int i = 1; i < 5; i++) {
    if (num_seq[i] < min) min = num_seq[i];
  }
  return(min);
}


int max(int num_seq[5]) {
  // Get the maximum from a numeric list.
  int max = num_seq[0];

  for (int i = 1; i < 5; i++) {
    if (num_seq[i] > max) max = num_seq[i];
  }
  return(max);
}


int get_max_id(int num_seq[5]) {
  // Get the id of the maximum value from a numeric list.
  int max = num_seq[0];
  int max_id = 0;

  for (int i = 1; i < 5; i++) {
    if (num_seq[i] > max) {
      max = num_seq[i];
      max_id = i;
    }
  }
  return(max_id);
}


int sum(int num_seq[5]) {
  // Sum the values in a numeric list.
  int sum = 0;

  for (int i = 0; i < 5; i++) sum += num_seq[i];

  return(sum);

}


void count_types_in_hand(int hand[5][2], int* types) {
  // Count the amount of distinct types in a poker hand.
  int j, i;
  *types = 0;

  for(i = 0; i < 5; i++) {

    for (j=0; j<i; j++){
      if (hand[i][1] == hand[j][1]) break;
    }

    if (i == j) *types += 1;
  }
}


void n_of_a_kind(int num_seq[5], int n_of_kind[2], int* n_of_kind_weight) {
  // Abstract function to calculate the occurence of repeat cards.
  int j = 0;
  *n_of_kind_weight = 0;
  int prev = num_seq[0];
  int count = 1;

  for (int i = 1; i < 5; i++)
  {
    if (num_seq[i] == prev) count++;
    else
    {
      if (count > 1) {
        n_of_kind[j++] = count;
        *n_of_kind_weight += count * prev;
      }
      prev = num_seq[i];
      count = 1;
    }
  }

  if (count > 1) {
    n_of_kind[j++] = count;
    *n_of_kind_weight += count * prev;
  }

}


bool is_hand_sequential(int num_seq[5]) {
  // Determin if a numeric sequence can be sequential.
  for (int i = 1; i < 5; i++) {
    if (num_seq[i-1] != (num_seq[i] - 1)) return(false);
  }

  return(true);

}


void get_hands(FILE* in_file, int l_hand[5][2], int r_hand[5][2]) {
  // Parse CodeEval problem set Poker Hands into a usable format.
  char parse_l_hand[5][3], parse_r_hand[5][3];

  for (int i = 0; i < 5; i++) {
    fscanf(in_file, "%s ", parse_l_hand[i]);
  }

  for (int i = 0; i < 4; i++) {
    fscanf(in_file, "%s ", parse_r_hand[i]);
  }
  fscanf(in_file, "%s\n", parse_r_hand[4]);

  convert_to_numeric(parse_l_hand, l_hand);
  convert_to_numeric(parse_r_hand, r_hand);

}


void convert_to_numeric(char parse_hand[5][3], int hand[5][2]) {
  // Encode string formated card values into numeric representation.
  for (int i = 0; i < 5; i++) {
    if (isdigit(parse_hand[i][0])) hand[i][0] = parse_hand[i][0] - '0' ;
    else hand[i][0] = name_to_num(parse_hand[i][0]);

    if (isdigit(parse_hand[i][1])) hand[i][1] = parse_hand[i][1] - '0' ;
    else hand[i][1] = name_to_num(parse_hand[i][1]);
  }
}


int name_to_num(char name) {
  // Mapping for string values of cards to numeric values.
  switch(name) {

    // Royal name converstion
    case 'T' : return(10);
    case 'J' : return(11);
    case 'Q' : return(12);
    case 'K' : return(13);
    case 'A' : return(14);

    // Type converstion
    case 'C' : return(0);
    case 'D' : return(1);
    case 'H' : return(2);
    case 'S' : return(3);
  }
}

int cmpfunc (const void * a, const void * b) {
  // Computation function for the qsort algorithm.
  return ( *(int*)a - *(int*)b );
}
